### ISSUE TITLE
Tasks should run `dependsOn` before hashing inputs

### ISSUE THREAD

**Issue by JavaScriptBach** [vercel/turborepo#8051]
### Verify canary release

- [X] I verified that the issue exists in the latest Turborepo canary release.

### Link to code that reproduces this issue

https://github.com/JavaScriptBach/turbo-caching-bug

### What package manager are you using / does the bug impact?

Yarn v2/v3/v4 (node_modules linker only)

### What operating system are you using?

Linux

### Which canary version will you have in your reproduction?

1.13.3-canary.4

### Describe the Bug

If I have a task A whose input is a generated file produced by another task B, I have to Turbo task A twice before it gets cached.

## What I think is happening

I think it's because on the first invocation:

1. Turbo hashes the input, but the generated file doesn't exist yet.
2. Turbo runs the dependent task, which produces the generated file
3. Turbo runs the original task and stores it under the now-stale hash.

On second invocation:
1. Turbo sees that the generated file now exists, so the hash is different from before, therefore it runs the original task again.

## Why I think it's a bug
In order to obtain correct caching behavior, I currently have to exclude all generated files from the input. This is non-intuitive because the generated files are conceptually inputs to my task. Furthermore, I've already told Turbo that it depends on my codegen task.

It would be nice for Turbo to handle this, perhaps by running all dependsOn tasks before hashing the inputs to the original task?

### Expected Behavior

Everything is cached after running Turbo once.

### To Reproduce

See the linked repo.

**Comment by peplin** [comment]
Here's an example of 3 sequential `turbo` runs from the linked repository. You can see how it takes 3 runs to get to `FULL TURBO`:

```
$ turbo my-test --summarize                                                                                                                     
• Running my-test
• Remote caching disabled
my-codegen: cache miss, executing fb830814bce3d882
my-codegen:
my-test: cache miss, executing 2ef6da69cc2b018d
my-test:

  Tasks:    2 successful, 2 total
 Cached:    0 cached, 2 total
   Time:    373ms
Summary:    /Users/peplin/dev/turbo-caching-bug/.turbo/runs/2fejXqPcSUhLBGZsQgZ3Flk77ml.json


$ node_modules/.bin/turbo my-test --summarize
• Running my-test
• Remote caching disabled
my-codegen: cache hit, replaying logs fb830814bce3d882
my-codegen:
my-test: cache miss, executing fcb34ccdb1cf4377
my-test:

  Tasks:    2 successful, 2 total
 Cached:    1 cached, 2 total
   Time:    242ms
Summary:    /Users/peplin/dev/turbo-caching-bug/.turbo/runs/2fejY73vyb4eLqszBlFWviHymzE.json


$ node_modules/.bin/turbo my-test --summarize
• Running my-test
• Remote caching disabled
my-codegen: cache hit (outputs already on disk), replaying logs fb830814bce3d882
my-codegen:
my-test: cache hit, replaying logs fcb34ccdb1cf4377
my-test:

  Tasks:    2 successful, 2 total
 Cached:    2 cached, 2 total
   Time:    73ms >>> FULL TURBO
Summary:    /Users/peplin/dev/turbo-caching-bug/.turbo/runs/2fejYAIPKlkQD5auUDsWA7FLt2M.json
```

Here are the 3 summary files:

[1.json](https://github.com/vercel/turbo/files/15135564/1.json)
[2.json](https://github.com/vercel/turbo/files/15135566/2.json)
[3.json](https://github.com/vercel/turbo/files/15135565/3.json)

1.json does not include `my-codegen.txt` in its inputs, so the cache hash is different than run 2. I would expect the first run to be a cache miss, but with a generated hash that matches the second run.

**Comment by weyert** [comment]
Yeah, I think I might have a similar problem but for the `codegen`-task is in the same package, e.g. using:

```json
{
  "$schema": "https://turbo.build/schema.json",
  "extends": ["//"],
  "pipeline": {
    "generate": {
      "outputMode": "new-only",
      "inputs": [
        "src/**/*.yml"
      ],
      "outputs": ["src/**/*", "!src/**/*.yml"],
      "cache": true
    },
    "build": {
      "outputMode": "new-only",
      "inputs": [
        "!src/**/*.yml"
      ],
      "outputs": ["lib/**"],
      "dependsOn": ["generate"],
      "cache": true
    }
  }
}
```

**Comment by mattico** [comment]
I also ran into this issue and made a repro of my own before I found yours: https://github.com/mattico/turborepo-repro

I can verify this issue still exists with version 2.0.15-canary.4.

**Comment by Leksat** [comment]
Just met this issue. Confirming that `2.1.2-canary.0` is affected.

**Comment by Leksat** [comment]
I created https://github.com/vercel/turborepo/pull/10253 with a test case for this issue. It's not perfect, but it works.

---

After doing a bit of research, I found that this issue is actually quite hard to solve.

Currently, Turborepo calculates input hashes in parallel for all tasks before executing them. This allows it to determine in advance which tasks will hit the cache and to create an efficient execution plan — which results in very fast builds.

However, it doesn't take into account that inputs and outputs from different tasks can intersect and change during execution.

How can this be solved without impacting performance? That’s a tough one. I guess only the Turborepo team can answer that, as it requires deep knowledge of the current architecture 😬

**Comment by david-wasser** [comment]
I just wanted to highlight that Turbo computing cache hits before executing tasks comes with a pretty high cost for any system that has multiple Turbo executions running concurrently. Particularly with something like continuous integration testing with a remote cache. It results in far fewer cache hits than if the cache is checked before each task is run.

For a slightly more concrete example, my org has a medium-sized monorepo, we use a merge queue to merge code into our main branch, and we use a Turbo remote cache. This merge queue builds and tests the first 20 PRs in the queue concurrently for higher throughput. However, due to Turbo computing cache hits before any tasks are executed, we're effectively guaranteed that we will duplicate all Turbo tasks necessary for the 19 PRs ahead of any particular PR in the queue because it will cache miss on all Turbo executions that are still in flight. This is particularly painful when the PRs touch totally different parts of the monorepo.

**Comment by anthonyshew** [comment]
This has now been implemented as a feature in #13043. Docs can be found here: https://turborepo.dev/docs/reference/configuration#turbo_jit

**Comment by david-wasser** [comment]
@solsson @anthonyshew thanks so much for implementing this after all this time!

Is there any chance this can just be turned on with a config for all `dependsOn` task hashes instead of for specific `inputs`? I never came back to this thread to update it because I assumed it was dead after 2 years of no responses from Vercel, but our main usecase for hashing after dependency tasks run is actually probably to enable [`isolatedDeclarations`](https://www.typescriptlang.org/tsconfig/#isolatedDeclarations). `isolatedDeclarations` can potentially give us massive build caching wins when typechecking/linting since any change that doesn't modify the exported signatures of a package can hit build cache for all transitive reverse-deps.

With `turbo` before your changes for this issue, even whitespace-only changes trigger all transitive reverse-deps to get cache misses. With `turbo` after your changes we can potentially avoid this. But all tasks must explicitly list all files they may transitively depend on from other packages as jit inputs if I'm understanding this correctly - which is pretty intractable in a non-trivially-sized monorepo.

**Comment by anthonyshew** [comment]
Pretty sure you're describing a different feature, depending on outputs of task dependencies (e.g. `$TURBO_OUTPUTS$`). That's different from this feature. This is for source code mutations mid-`turbo`.

(Distributed de-duping from your prior comment is also a different feature, and heavily platform dependent. Not something the `turbo` client can properly implement.)

**Comment by david-wasser** [comment]
> Pretty sure you're describing a different feature, depending on outputs of task dependencies (e.g. `$TURBO_OUTPUTS$`). That's different from this feature.

I don't think it is a different feature? Depending on outputs of task dependencies is exactly what's being done in the original post if you look at [the repro repo's `turbo.json`](https://github.com/JavaScriptBach/turbo-caching-bug/blob/main/turbo.json). `my-test` depends on the output of `my-codegen`.

I also wasn't really referring to some kind of remote build execution deduping in my other comment. I recognize that's not feasible with `turbo`. We use [ducktors/turborepo-remote-cache](https://github.com/ducktors/turborepo-remote-cache) for remote task caching currently. If the `dependsOn` hash was computed after the `dependsOn` tasks completed, that would resolve my other comment. The root issue to both of these problems is that all task hashes are computed before building any of them.

Also, to provide a little bit of context - I currently work at the company @JavaScriptBach and @peplin did when they originally posted this. For what it's worth, we have worked around the specific issue of `inputs` that're `outputs` of other tasks causing cache misses by giving explicit guidance internally that it should always be considered a bug to declare an `input` that is an `output` of a `dependsOn`. Instead, just declare the `dependsOn` without any `inputs` from that task.

**Comment by anthonyshew** [comment]
There's definitely some overlap in the two use cases, and the difference is subtle. The repro for the Issue could be explained as either one, and I chose to close it on the one I felt was nearer. The good news is I'm going to implement the solution for both, so both interpretations of the Issue will be covered!

In the case provided, the `outputs` being directly equal to `inputs` is something of an accident (in the way I read it). What's being asked for is to be able to dynamically capture source code changes into hashing inputs after tasks have been started. Semantically, the tasks that produced those changes don't matter. In fact, these files could have come from anywhere, if you're creative enough.

Another example of this case is a typical "format, then lint" pipeline. `format` won't have outputs, but you want to cache `lint` when the source code input to the hash will be the same. 

There's also another case, which is the one you're describing. `isolatedDeclarations` is a great example. You may want to semantically connect the `outputs` of a task to the `inputs` of its dependents. That's a *much* different use case. You want to specifically tie the `inputs` of a dependent to the `outputs` of its task dependency(ies).

It sounds like you were looking for the second one, but I did the first one, which I still feel like the Issue description describes more closely. But, rest assured, I'm going to figure out the second use case, as well.
